src/ds++/FMA.hh

```diff
@@ -26,41 +44,129 @@
+/*!
+ * If \c FMA_FIND_DIFFS is defined, then hijack the whole FMA operation.
+ * Instead, perform the operation both ways and compare the result.  If the
+ * result is significantly different, throw an exception.  This is useful for
+ * finding operations that are sensitive to roundoff.
+ */
+#define FMA(a, b, c)                                                           \
+  rtt_dsxx::fma_with_diagnostics((a), (b), (c), __FILE__, __LINE__, true)
+#define FMA_ACCURATE(a, b, c)                                                  \
+  rtt_dsxx::fma_with_diagnostics((a), (b), (c), __FILE__, __LINE__, false)
+#else
+
+//----------------------------------------------------------------------------//
+//! HAVE_HARDWARE_FMA is set by config/platform_checks.cmake.
 #ifdef HAVE_HARDWARE_FMA
 
-/* This will be the default for new machines. Both versions should be equally
- * accurate, choose whatever is faster. FP_FAST_FMA is set by the compiler
+//----------------------------------------------------------------------------//
+/*!
+ * This will be the default for new machines. Both versions should be equally
+ * accurate, choose whatever is faster. \c FP_FAST_FMA is set by the compiler
  *
  * \ref http://en.cppreference.com/w/c/numeric/math/fma
+ *
+ * Working on LANL systems, it seems that \c FP_FAST_FMA is not set by the Intel
+ * compiler, so Intel will always pick the 2nd option.
+ *
+ * The intent of the \c FMA_ACCURATE version is to pick the more accurate
+ * multiply-plus-add operation. When hardware FMA is available, we just just the
+ * normal 'fast' version.
  */
 #ifdef FP_FAST_FMA
-#define FMA(a, b, c) fma(a, b, c)
+#define FMA(a, b, c) fma((a), (b), (c))
+#define FMA_ACCURATE(a, b, c) fma((a), (b), (c))
 #else
 #define FMA(a, b, c) ((a) * (b) + c)
+#define FMA_ACURATE(a, b, c) ((a) * (b) + c)
 #endif
 
 #else /* HAVE_HARDWARE_FMA is false */
 
-/*! For older hardware that does not support FMA natively, provide a macro that
- *  will call an infinite precision fma function so that numerical
- *  reproducibility is enhanced (if FP_ACCURATE_FMA is defined). This allows
- *  solutions generated on older machines to match solutions generated on newer
- *  machines. */
-#ifndef FP_ACCURATE_FMA
-#error "Must set FP_ACCURATE_FMA to 0 or 1."
-#endif
-
-#if FP_ACCURATE_FMA > 0
-#define FMA(a, b, c) fma(a, b, c)
-#else
+//----------------------------------------------------------------------------//
+/*!
+ * For older hardware that does \b not support FMA natively, provide a macro
+ * that will call an infinite precision fma function so that numerical
+ * reproducibility is enhanced This allows solutions generated on older machines
+ * to match solutions generated on newer machines.
+ *
+ * Operations that are known to be very sensitive to round-off error should use
+ * \c FMA_ACCURATE even though it is much slower (2-3x slower).
+ */
 #define FMA(a, b, c) ((a) * (b) + c)
-#endif
+#define FMA_ACCURATE(a, b, c) fma((a), (b), (c))
```

Comment:
I see, this is where the FMA_ACCURATE and FMA are different.

Reply:
Is this going to bite us if we fail to pickup the native FMA on the platform?

Reply:
* Right
* We will deal with that when we encounter it.  As always, the problem is reproducibility of solutions between platforms and compilers and compiler options.  We could always choose the 'fastest', but then our test setup is worthless.

Reply:
I'm sure we will be revisiting this whole issue later this year due to porting to GPU and ARM.